File: Python/2-TwoPointers/Problems/1-ValidPalindrome/ValidPalindromePro.py
# Advanced Valid Palindrome Implementations
# Real-World Applications of Palindrome Checking

import re  # Import the regular expression module for string manipulation
from collections import deque  # Import the deque class for efficient left/right popping

# Approach 1: Real-Time Text Processing (Efficient Two Pointers)
# Used in text validation for secure inputs in chat applications
def is_palindrome_real_time(s):
    """
    Checks if a string is a palindrome in real-time, ignoring non-alphanumeric characters.

    This approach is efficient because it uses two pointers, one starting from the
    beginning of the string and the other from the end, moving towards the center.
    It's suitable for applications where speed is critical, such as validating
    user input in chat applications as it's being typed.

    Args:
        s: The string to check.

    Returns:
        True if the string is a palindrome, False otherwise.
    """
    left, right = 0, len(s) - 1  # Initialize left and right pointers
    while left < right:  # Continue as long as the left pointer is before the right pointer
        while left < right and not s[left].isalnum():
            left += 1  # Move left pointer to the next alphanumeric character
        while left < right and not s[right].isalnum():
            right -= 1  # Move right pointer to the previous alphanumeric character
        if s[left].lower() != s[right].lower():
            return False  # If characters don't match, it's not a palindrome
        left += 1  # Move left pointer one step forward
        right -= 1  # Move right pointer one step backward
    return True  # If the loop finishes without finding mismatches, it's a palindrome

# Approach 2: Search Engine Optimization (Ignoring Stopwords)
# Used in SEO ranking algorithms to compare search queries
def is_palindrome_seo(s):
    """
    Checks if a string is a palindrome, ignoring common stopwords.

    This approach is useful in SEO for comparing search queries, where certain
    words like "a", "the", and "is" can be ignored to focus on the more
    meaningful parts of the query.  This helps in ranking similar queries.

    Args:
        s: The string to check.

    Returns:
        True if the string is a palindrome (after removing stopwords), False otherwise.
    """
    stopwords = {"a", "the", "is", "in", "at", "of", "and"}  # Define a set of stopwords
    words = re.findall(r"\w+", s.lower())  # Extract words, convert to lowercase
    filtered = [word for word in words if word not in stopwords]  # Remove stopwords
    return filtered == filtered[::-1]  # Check if the filtered list is equal to its reverse

# Approach 3: Genetic Sequencing (DNA Palindrome Check)
# Used in bioinformatics for DNA pattern recognition
def is_palindrome_dna(sequence):
    """
    Checks if a DNA sequence is a palindrome (reverse complement).

    In genetics, palindromic sequences are important for processes like DNA
    replication and gene regulation. This function checks if a DNA sequence
    is equal to its reverse complement.

    Args:
        sequence: The DNA sequence string (e.g., "ATCGCGAT").

    Returns:
        True if the DNA sequence is a palindrome, False otherwise.
    """
    complement = {"A": "T", "T": "A", "C": "G", "G": "C"}  # Define base complements
    rev_comp = ''.join(complement.get(base, base) for base in reversed(sequence))  # Generate the reverse complement
    return sequence.upper() == rev_comp #check if the uppercase version of sequence is equal to the reverse complement

# Approach 4: Data Validation in Databases (Using SQL Normalization)
# Used in database systems to validate and compare normalized text inputs
def is_palindrome_database(s):
    """
    Checks if a string is a palindrome after normalizing it for database storage.

    Databases often need to validate data to ensure consistency.  This function
    normalizes a string by removing non-alphanumeric characters and converting
    it to lowercase, which is a common step before comparing strings in a database.

    Args:
        s: The string to check.

    Returns:
        True if the normalized string is a palindrome, False otherwise.
    """
    normalized = re.sub(r'[^a-zA-Z0-9]', '', s).lower()  # Normalize the string
    return normalized == normalized[::-1]  # Check if the normalized string is a palindrome

# Approach 5: AI Chatbot Responses (Efficient with Deque)
# Used in AI/NLP systems to recognize palindromic queries efficiently
def is_palindrome_chatbot(s):
    """
    Checks if a string is a palindrome using a deque (double-ended queue).

    A deque is more efficient than a regular list for this purpose because
    removing elements from both ends is fast.  This is useful in AI chatbots
    for quickly processing user input to understand if a query is palindromic.

    Args:
        s: The string to check.

    Returns:
        True if the string is a palindrome, False otherwise.
    """
    d = deque()  # Initialize a deque
    for char in s:
        if char.isalnum():
            d.append(char.lower())  # Add alphanumeric characters to the deque
    while len(d) > 1:
        if d.popleft() != d.pop():
            return False  # Compare characters from both ends
    return True  # If the deque is empty or has one element, it's a palindrome

# Approach 6: Cryptographic Security (Reverse Hash Matching)
# Used in cryptographic hash functions to validate integrity
def is_palindrome_crypto(s):
    """
    Checks if a string is a palindrome by comparing the hashes of the string
    and its reverse.

    In cryptography, hashing is used to ensure data integrity.  This function
    checks if a string is a palindrome by comparing the MD5 hash of the
    normalized string with the MD5 hash of its reverse.

    Args:
        s: The string to check.

    Returns:
        True if the hashes of the string and its reverse match, False otherwise.
    """
    import hashlib  # Import the hashlib module
    normalized = re.sub(r'[^a-zA-Z0-9]', '', s).lower()  # Normalize the string
    forward_hash = hashlib.md5(normalized.encode()).hexdigest()  # Hash the forward string
    reverse_hash = hashlib.md5(normalized[::-1].encode()).hexdigest()  # Hash the reversed string
    return forward_hash == reverse_hash  # Compare the hashes

# Test Cases
test_cases = [
    "A man, a plan, a canal: Panama",  # Classic palindrome
    "race a car",  # Not a palindrome
    "ATCGCGAT",  # DNA sequence (not palindromic)
    "Was it a car or a cat I saw",  # Sentence palindrome
    "Madam In Eden, I’m Adam",  # Complex case
]

# Iterate through the test cases and print the results of each palindrome check.
for i, test in enumerate(test_cases, 1):
    print(f"Test Case {i}: {test}")
    print("Real-Time Text Processing:", is_palindrome_real_time(test))
    print("SEO Optimization:", is_palindrome_seo(test))
    print("DNA Palindrome Check:", is_palindrome_dna(test.replace(' ', '').upper()))
    print("Database Validation:", is_palindrome_database(test))
    print("Chatbot Processing:", is_palindrome_chatbot(test))
    print("Cryptographic Security:", is_palindrome_crypto(test))
    print()

File: C++/2-TwoPointers/Problems/1-ValidPalindrome/ValidPalindromePro.js
/**
 * # Two Pointers: Valid Palindrome
 *
 * A palindrome is a word, phrase, number, or other sequence of characters which reads the same backward as forward.
 *
 * Given a string `s`, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
 *
 * ## Problem Constraints
 *
 * -   `1 <= s.length <= 2 * 10^5`
 * -   `s` consists only of printable ASCII characters.
 *
 * ## Examples
 *
 * ### Example 1
 *
 * ```
 * Input: s = "A man, a plan, a canal: Panama"
 * Output: true
 * Explanation: "amanaplanacanalpanama" is a palindrome.
 * ```
 *
 * ### Example 2
 *
 * ```
 * Input: s = "race a car"
 * Output: false
 * Explanation: "raceacar" is not a palindrome.
 * ```
 *
 * ### Example 3
 *
 * ```
 * Input: s = " "
 * Output: true
 * Explanation: An empty string is considered a palindrome.
 * ```
 */

/**
 * ---
 * 1.  **Two Pointers (Optimized)**
 *
 * -   **Approach:**
 * -   Initialize two pointers, `left` at the beginning and `right` at the end of the string.
 * -   Iterate while `left < right`.
 * -   In each iteration, move `left` to the right until it points to an alphanumeric character.
 * -   Similarly, move `right` to the left until it points to an alphanumeric character.
 * -   Compare the characters at `left` and `right` (case-insensitive). If they are not equal, the string is not a palindrome.
 * -   If the characters are equal, move `left` to the right and `right` to the left.
 * -   If the loop finishes without finding any mismatched characters, the string is a palindrome.
 * -   **Real-World Project:**
 * -   Text editors or IDEs for checking if a user's input is a valid palindrome for specific features (e.g., variable names, special syntax).
 * -   **Complexity Analysis:**
 * -   Time Complexity: O(n), where n is the length of the string.
 * -   Space Complexity: O(1), constant extra space.
 * -   **Explanation:**
 * This approach efficiently skips non-alphanumeric characters, directly comparing only the relevant characters.  It's optimized for speed and uses minimal memory.
 */
function isPalindromeTwoPointers(s) { // Removed the type annotation ": string"
    let left = 0;
    let right = s.length - 1;

    while (left < right) {
        // Skip non-alphanumeric characters from the left
        while (left < right && !isAlphaNumeric(s[left])) {
            left++;
        }
        // Skip non-alphanumeric characters from the right
        while (left < right && !isAlphaNumeric(s[right])) {
            right--;
        }

        // Compare characters (case-insensitive)
        if (left < right && s[left].toLowerCase() !== s[right].toLowerCase()) {
            return false;
        }

        // Move pointers
        left++;
        right--;
    }

    return true;
}

/**
 * Helper function to check if a character is alphanumeric.
 */
function isAlphaNumeric(char) { // Removed the type annotation ": string"
    const code = char.charCodeAt(0);
    return (
        (code >= 48 && code <= 57) || // 0-9
        (code >= 65 && code <= 90) || // A-Z
        (code >= 97 && code <= 122)   // a-z
    );
}

/**
 * ---
 * 2.  **Two Pointers (Simplified)**
 * -   **Approach:**
 * -   Similar to the optimized two-pointer approach, but combines the character validation and comparison into a single, more concise loop.
 * -   It initializes two pointers, `left` and `right`, at the start and end of the string.
 * -   The loop continues as long as `left < right`.
 * -   Inside the loop, it checks if the characters at `left` and `right` are alphanumeric. If not, it moves the respective pointer and continues.
 * -   If both characters are alphanumeric, it compares them (case-insensitively). If they don't match, it returns `false`.
 * -   If they match, it moves both pointers towards the center.
 * -   If the loop finishes without finding mismatches, it returns `true`.
 * -   **Real-World Project:**
 * -   Used in data validation within web applications to ensure user-provided strings (e.g., usernames, IDs) adhere to palindrome rules, enhancing data integrity.
 * -   **Complexity Analysis:**
 * -   Time Complexity: O(n), where n is the length of the string.
 * -   Space Complexity: O(1), constant extra space.
 */
function isPalindromeTwoPointersSimple(s) { // Removed the type annotation ": string"
    let left = 0;
    let right = s.length - 1;

    while (left < right) {
        const leftChar = s[left];
        const rightChar = s[right];

        if (!isAlphaNumeric(leftChar)) {
            left++;
        } else if (!isAlphaNumeric(rightChar)) {
            right--;
        } else if (leftChar.toLowerCase() !== rightChar.toLowerCase()) {
            return false;
        } else {
            left++;
            right--;
        }
    }
    return true;
}

/**
 * ---
 * 3.  **Using Built-in Functions (Filter and Reverse)**
 * -   **Approach:**
 * -   Filter the input string to keep only alphanumeric characters.
 * -   Convert the filtered string to lowercase.
 * -   Reverse the filtered string.
* -  Compare the filtered string with its reversed version.
 * -   **Real-World Project:**
 * -   Used in natural language processing (NLP) tasks for text normalization before palindrome checks, ensuring that only relevant characters are considered.
 * -   **Complexity Analysis:**
 * -   Time Complexity: O(n), where n is the length of the string.  Filtering and reversing are linear operations.
 * -   Space Complexity: O(n), as a new string is created to store the filtered characters and the reversed string.
 */
function isPalindromeFilterReverse(s) { // Removed the type annotation ": string"
    const filteredChars = s.toLowerCase().split('').filter(isAlphaNumeric);
    const reversedChars = [...filteredChars].reverse(); //spread syntax to create new array.
    return filteredChars.join('') === reversedChars.join('');
}

/**
 * ---
 * 4.  **Using Regular Expression and Reverse**
 * -   **Approach:**
 * -   Use a regular expression to remove all non-alphanumeric characters from the string.
 * -   Convert the cleaned string to lowercase.
 * -   Reverse the cleaned string.
 * -   Compare the cleaned string with its reversed version.
 * -   **Real-World Project:**
 * -   Used in data cleaning processes to validate if user inputs (e.g., names, addresses) are palindromes after removing irrelevant symbols or formatting.
 * -   **Complexity Analysis:**
 * -   Time Complexity: O(n), where n is the length of the string.  Regular expression matching and reversing are linear.
 * -   Space Complexity: O(n), for the cleaned string and the reversed string.
 */
function isPalindromeRegExp(s) { // Removed the type annotation ": string"
    const cleanedString = s.replace(/[^a-zA-Z0-9]/g, '').toLowerCase();
    const reversedString = cleanedString.split('').reverse().join('');
    return cleanedString === reversedString;
}

/**
 * ---
 * 5.  **Recursive Approach**
 * -   **Approach:**
 * - Define a recursive function that takes the string and two pointers (left, right) as arguments.
 * - In each recursive call:
* - If left >= right, the substring is a palindrome (base case), return true
 * -  Move left and right inwards, skipping non-alphanumeric characters.
 * -  If the characters at left and right are different, return false.
 * -  Otherwise, recursively call the function with the updated left and right pointers.

 * -   **Real-World Project:**
 * -   While less common for simple palindrome checks, recursion can be adapted for more complex, nested data structures where palindrome-like properties need to be validated.
 * -   **Complexity Analysis:**
 * -   Time Complexity: O(n), where n is the length of the string.
 * -   Space Complexity: O(n) in the worst case, due to the call stack in recursion.
 */

function isPalindromeRecursive(s) { // Removed the type annotation ": string"
  function solve(str, left, right) { // Removed type annotations ": string", ": number", ": number"
    if (left >= right) {
      return true;
    }

    while (left < right && !isAlphaNumeric(str[left])) {
      left++;
    }
    while (left < right && !isAlphaNumeric(str[right])) {
      right--;
    }

    if (left < right && str[left].toLowerCase() !== str[right].toLowerCase()) {
      return false;
    }

    return solve(str, left + 1, right - 1);
  }
  return solve(s, 0, s.length - 1);
}

/**
 * 6. **Using Array.every()**
 * - **Approach:**
 * - Filter the string to keep only alphanumeric characters and convert it to lowercase.
 * - Use the `every()` method to check if each character in the filtered string is equal to its corresponding character from the end of the string.
 * - **Real-world project:**
 * - Useful in data validation scenarios where you need to check a property for all elements in a list, and that property is whether the element is a palindrome.
 * - **Complexity Analysis:**
 * - Time complexity: O(n), where n is the length of the string.
 * - Space complexity: O(n), to store the filtered string.
 */
function isPalindromeEvery(s) { // Removed the type annotation ": string"
    const cleanStr = s.toLowerCase().split('').filter(isAlphaNumeric);
    return cleanStr.every((char, i) => char === cleanStr[cleanStr.length - 1 - i]);
}

// ---
// ## Test Cases
//
// The following test cases cover various scenarios, including empty strings, strings with non-alphanumeric characters, and palindromes/non-palindromes of different lengths.
//
function runTestCases() {
    const testCases = [
        { input: "A man, a plan, a canal: Panama", expected: true },
        { input: "race a car", expected: false },
        { input: " ", expected: true },
        { input: "Madam, I'm Adam!", expected: true },
        { input: "121", expected: true },
        { input: "123", expected: false },
        { input: "0P", expected: false},
        { input: "a.", expected: true},
        { input: "", expected: true }, // Empty string
        { input: "ab", expected: false },
        { input: "aba", expected: true },
        { input: "abba", expected: true },
        { input: "Aba", expected: true },  //Mixed case
    ];

    const implementations = [
        { name: "Two Pointers (Optimized)", func: isPalindromeTwoPointers },
        { name: "Two Pointers (Simplified)", func: isPalindromeTwoPointersSimple },
        { name: "Filter and Reverse", func: isPalindromeFilterReverse },
        { name: "RegExp and Reverse", func: isPalindromeRegExp },
        { name: "Recursive", func: isPalindromeRecursive },
        { name: "Using Array.every()", func: isPalindromeEvery},
    ];

    testCases.forEach((testCase, index) => {
        console.log(`\n--- Test Case ${index + 1} ---`);
        console.log(`Input: "${testCase.input}"`);
        console.log(`Expected Output: ${testCase.expected}`);

        implementations.forEach(implementation => {
            const actual = implementation.func(testCase.input);
            console.log(`${implementation.name}: ${actual === testCase.expected ? '✅ Passed' : '❌ Failed'}, Output: ${actual}`);
        });
    });
}

runTestCases();
